Summon Monsters

Alium58 opened the pull request:
# Description

Uses monster summoning function from #1176 to actually summon monsters. Put quite a few options in there since combat locket provides 3 summons a day on it's own.

- Summoning mostly happens in its own function that gets called fairly high in the task list
  - This allows for summons to happen when we are ready for them instead of whenever we attempt their associated quest
- Removed LX_Faxing as that is taken care of in summoning code now
- Removed original monster summoning function that was never implemented. Didn't know this one existed when I made #1176 
- ~~Added Ghost as a backup camera target~~
- Removed bullet bill banish. Missed in recent 8bit realm update #1241 

## How Has This Been Tested?

Multiple normal seal clubber runs. Haven't tried any HC runs yet. Don't plan to unless someone would like to see how it runs.

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have based my pull request against the [main branch](https://github.com/Loathing-Associates-Scripting-Society/autoscend/tree/main) or have a good reason not to.

quarklikeadork commented:
looking only at summon choices,

L4_batCave uses 1 Sonar-In-A-Biscuit at a time, does it get opportunity to use it for last wall before screambat will get summoned?

Ninja Snowman Assassin comes without a minimum quest state so looks like it will be target for summon earlier than the conditions to copy it, back up wants level 8

smut orc pervert also comes with no minimum quest state or check for possessed Smut Orc Keepsake Box so it will be too eager to summon as many as it can at run start?

Craggy bartender could be another good target in standard if decent booze is scarce? but that could be hard to determine

Alium58 commented:
Added check for sonar-in-a-biscuit before summoning screambat

Added level 8 check for NSA

Smut orc quest status is included in the lumberCount() and fastenerCount() functions. But doesn't account for keepsake boxes on hand. Added check to summon only if no keepsake boxes are on hand. More for future proofing as I believe only cargo shorts can summon this guy and it is 1/day.

Preadv isn't called for summoning. Added call to maximizer for ghosts to get more +item

Agree craggy might be a good option. Perhaps simulate drinking and if next drink is < 4 adv/liver and we can drink craggy drops, then summon?

Alium58 commented:
Going to do at least one more run to test recent changes as well as non-standard summoning. Like faxing. Will update when I complete a non-standard run

quarklikeadork commented on RELEASE/scripts/autoscend/auto_util.ash:
```diff
@@ -1961,9 +1797,83 @@ boolean auto_deleteMail(kmailObject msg)
+		if(summonMonster($monster[lobsterfrogman])) return true;
+	}
+
+	if(auto_is_valid($item[Smut Orc Keepsake Box]) && item_amount($item[Smut Orc Keepsake Box]) == 0 && (lumberCount() < 30 || fastenerCount() < 30))
```
suggest adding a condition like
>(my_level() >= 9 || my_adventures() < 7+ auto_advToReserve())

to leave an opportunity to get the more valuable level 8 mountain man first if cargo summon is its only available source?
!(auto_have_skill($skill[Stuffed Mortar Shell]) && auto_have_skill($skill[Spirit of Peppermint])) could be added as another rough condition for not waiting for mountain man, ores being more valuable than keepsake box is assuming blech house every 3 fights

Alium58 replied:
Good point. Added the level check both for this reason but also so we know we can beat the monster as no yellow ray is used

quarklikeadork commented on RELEASE/scripts/autoscend/quests/level_11.ash:
```diff
@@ -1344,6 +1313,15 @@ boolean L11_unlockHiddenCity()
+		if(item_amount($item[Stone Wool]) == 0 && have_effect($effect[Stone-Faced]) == 0)
+		{
+			//attempt to summon before using a clover
+			handleFamiliar("item");
```
similar to ghost a maximizer call for item 400max should go here?

Alium58 replied:
Good call, did it

quarklikeadork commented on RELEASE/scripts/autoscend/quests/level_11.ash:
```diff
@@ -1344,6 +1313,15 @@ boolean L11_unlockHiddenCity()
+		if(item_amount($item[Stone Wool]) == 0 && have_effect($effect[Stone-Faced]) == 0)
+		{
+			//attempt to summon before using a clover
+			handleFamiliar("item");
```
this would run whenever wanting to summon the monster even if it won't happen, and handleFamiliar would lock the regular choice to item familiar,
so annoyingly handling familiar and equipment needs to go either in the individual summon methods or under a new function that checks if the summon will be available

Alium58 replied:
created new function `canSummonMonster` to check if we can summon prior to adding to maximizer / picking a fam

quarklikeadork commented on RELEASE/scripts/autoscend/auto_util.ash:
```diff
@@ -1961,9 +1797,83 @@ boolean auto_deleteMail(kmailObject msg)
+	// summon ghost for white pixels if at tower door and still need digital key
+	if(needDigitalKey() && !auto_hasPowerfulGlove() && internalQuestStatus("questL13Final") == 5)
+	{
+		handleFamiliar("item");
+		addToMaximize("20 item");
+		equipMaximizedGear();
```
this would run whenever wanting to summon the monster, same issues as for the stone wool summon

Alium58 replied:
removed ghost summoning

Alium58 commented on RELEASE/scripts/autoscend/auto_util.ash:
```diff
@@ -1961,9 +1797,83 @@ boolean auto_deleteMail(kmailObject msg)
+	if(needDigitalKey() && !auto_hasPowerfulGlove() && internalQuestStatus("questL13Final") == 5)
+	{
+		handleFamiliar("item");
+		addToMaximize("20 item");
```
Removed ghost summoning due to 8-bit realm changes

Alium58 commented on RELEASE/scripts/autoscend/auto_util.ash:
```diff
@@ -1961,9 +1797,83 @@ boolean auto_deleteMail(kmailObject msg)
+		// attempt to use calculate the universe
+		if(summonMonster($monster[War Frat 151st Infantryman])) return true;
+		// attempt to summon other sources of outfit
+		if(summonMonster($monster[War Frat Mobile Grill Unit])) return true;
```
No following. Can you explain with different words please

Alium58 replied:
Added level req for summoning so we don't do it too soon and miss out on a summon that would help the current day. Level req is 9 instead off 12 as that is when the script gets the pre-war outfit. Best to skip that and summon straight for the war outfit

Alium58 commented on RELEASE/scripts/autoscend/auto_util.ash:
```diff
@@ -1961,9 +1797,83 @@ boolean auto_deleteMail(kmailObject msg)
+	item oreGoal = to_item(get_property("trapperOre"));
+	if(internalQuestStatus("questL08Trapper") < 2 && oreGoal != $item[none] && item_amount(oreGoal) < 2 && adjustForYellowRayIfPossible())
+	{
+		if(summonMonster($monster[mountain man])) return true;
```
Agree it could happen but imo risk is low enough to not write extra code for it. Any insta kill yellow ray will get used prior to envy. If a user says it happens to them, then we can work on it

Alium58 commented:
Alrighty finally got back on this PR. Resolved all comments to date, thank you for them!

Going to do a another standard and then non-standard run to test as much as I can. Current standard HC TT run is going well except didn't summon NSA. Will look into that as well.

Alium58 commented:
Ok have ran a few HC standard runs and in a HC Ed run now to test unrestricted. Looking good and ready for final review/approval imo

Alium58 commented:
HC Ed run finished and went well. All done testing now
